zephyr/lib/regions_mm.c

```diff
@@ -1,54 +1,574 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ * Copyright(c) 2023 Intel Corporation. All rights reserved.
  *
  * Author: Jakub Dabek <jakub.dabek@intel.com>
  */
 
 #include <sof/lib/regions_mm.h>
 
-struct virtual_memory_heap
-	vm_heaps[CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT];
+/**
+ * @brief Initialize new heap
+ *
+ * Initializes new heap based on region that heap is to be assigned to
+ * and config.
+ * @param cfg Configuration structure with block structure for the heap
+ * @param memory_region_attribute A virtual region identifier
+ * @param span Flag deciding if heap can do contiguous allocation.
+ *
+ * @retval ptr to a new heap if successful.
+ * @retval NULL on creation failure.
+ */
+struct virtual_memory_heap *vmh_init_heap(const struct heap_config *cfg,
+					  int memory_region_attribute,
+					  bool span)
+{
+	uint32_t core_id = cpu_get_id();
+	struct sys_mm_drv_region *virtual_memory_regions =
+	    (struct sys_mm_drv_region *)sys_mm_drv_query_memory_regions();
+
+	/* Check if we haven't created heap for this region already */
+	if (vmh_get_heap_by_attribute(memory_region_attribute, core_id) != NULL)
+		return NULL;
+
+	struct virtual_memory_heap *new_heap =
+	    rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+		    sizeof(struct virtual_memory_heap));
```

Comment:
You can use `sizeof(*new_heap)` here

Reply:
I am little split on this. On one hand less code on the other it is clearer for me in this form. I would like to leave it as is if there is no really good reason for change.

Reply:
If the type of the variable changes, you can have an inconsistency here that can lead to undefined behaviour. The other style with sizeof(*new_heap) averts that danger.